### PR TITLE
Cover Finder

### PR DESCRIPTION
**Pull request by mariotaku**
## Description
This PR adds functionality to search game cover image, collected from IGDB.com.

### Screenshot
image

## Type of Change

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)

## Checklist

- [ ] My code follows the style guidelines of this project
- [ ] I have performed a self-review of my own code
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have added or updated the docstring/documentation-blocks for new or existing methods/components

### CODE REVIEW

File: sunshine/confighttp.cpp
Comment (TheElixZammuto) [inline sunshine/confighttp.cpp:460]: Probably we should check here if the URL provided in input is actually a trusted one (as if, from IGDB or similar)

File: src_assets/common/assets/web/apps.html
Comment (TheElixZammuto) [inline src_assets/common/assets/web/apps.html:323]: I can probably assume this is an instance of https://github.com/mariotaku/gcdb , am I right?

**Review by TheElixZammuto — COMMENTED** [review]
Unfortunately I'm not able to test this yet, but I have a couple feedback while skimming the code

File: src_assets/common/assets/web/apps.html
Comment (ReenigneArcher) [reply src_assets/common/assets/web/apps.html:323]: > I can probably assume this is an instance of https://github.com/mariotaku/gcdb , am I right?

I believe that is a correct assumption.

These two items appear to be equal.
https://mariotaku.org/gcdb/buckets/cc.json
https://github.com/mariotaku/gcdb/blob/main/docs/buckets/cc.json

I don't know that the first site is pulling directly from github though. I didn't see any call to github when inspecting the site. Maybe it is cloned locally on the server running the site?

Seems that the buck is the first 2 characters of the game title.

Comment (mariotaku) [reply src_assets/common/assets/web/apps.html:323]: The site is GH pages with custom domain. I'd like to ask if I could transfer this repo to host under SunshineStream.

Comment (ReenigneArcher) [reply src_assets/common/assets/web/apps.html:323]: You can transfer it... I assume I'll need to setup a workflow to update it on a schedule?

Comment (mariotaku) [reply src_assets/common/assets/web/apps.html:323]: Yes. I'll add this part and transfer the repo.

Comment (ReenigneArcher) [reply src_assets/common/assets/web/apps.html:323]: Okay, I can add secrets if needed... I assume I'll need a few for igdb api.

Comment (mariotaku) [reply src_assets/common/assets/web/apps.html:323]: I'm ready for transfer. Could you grant me permission to create public repo under SunshineStream org?

Comment (ReenigneArcher) [reply src_assets/common/assets/web/apps.html:323]: @mariotaku I just imported it (much faster)... https://github.com/SunshineStream/db

https://sunshinestream.github.io/db/

File: sunshine/confighttp.cpp
Comment (mariotaku) [reply sunshine/confighttp.cpp:468]: Good suggestion! Now only `images.igdb.com` is allowed. Also, `key` will be percent encoded to prevent special chars.

File: src_assets/common/assets/web/apps.html
Comment (mariotaku) [reply src_assets/common/assets/web/apps.html]: @ReenigneArcher Thanks. I'll submit to that repo for updates if needed.

**Comment by ReenigneArcher** [comment]
Flatpak will fail in the build. I have a fix, but am waiting until after I merge #204 and #214 .

**Comment by ReenigneArcher** [comment]
Unable to find `curl` on linux.

**Comment by ReenigneArcher** [comment]
We should probably add curl here...

https://github.com/SunshineStream/Sunshine/blob/nightly/CMakeLists.txt#L607-L608

I usually use this to get the package names: https://pkgs.org/search/?q=libcurl

**Comment by mariotaku** [comment]
This should be good. Except for flatpak configuration because I'm not sure how to properly ship libcurl.

**Comment by ReenigneArcher** [comment]
@mariotaku Can you update your branch with latest changes from `nightly`? This "should" fix the flatpak build failing.

Although it won't include curl. @istori1 can probably take a look at adding curl to the flatpak.

**Comment by ReenigneArcher** [comment]
@mariotaku I have a question. Is the sql database (https://github.com/SunshineStream/db/blob/gh-pages/gcdb.sqlite3) used by sunshine or at all? I didn't see any reference to it, although I might have missed it.

I'm thinking it's not used currently, and just put there for a possible future scenario.

**Comment by ReenigneArcher** [comment]
I haven't gotten around to activating the actions quite yet. It's on my to do list.

Okay, so you are creating the json file in the buckets after the sql database?

You can extract everything directly to json from igdb. https://github.com/twitchtv/igdb-api-python

**Comment by mariotaku** [comment]
By downloading to database, it can fetch recently updated items only. So it will save a huge amount of time.

**Comment by mariotaku** [comment]
In other words, right now this repo is doing differential updates, without fetching whole IGDB everyday.

File: packaging/linux/flatpak/com.github.sunshinestream.sunshine.yml
Comment (ReenigneArcher) [inline packaging/linux/flatpak/com.github.sunshinestream.sunshine.yml:70]: I was informed that it is not necessary to build curl as `libcurl.so` is already included in the flatpak runtime.

**Review by ReenigneArcher — CHANGES_REQUESTED** [review]

File: packaging/linux/flatpak/com.github.sunshinestream.sunshine.yml
Comment (mariotaku) [reply packaging/linux/flatpak/com.github.sunshinestream.sunshine.yml:70]: Excellent. I'll revert to last commit.

**Comment by ReenigneArcher** [comment]
@mariotaku would you mind updating your PR branch?

**Review by ReenigneArcher — APPROVED** [review]

**Comment by ReenigneArcher** [comment]
Testing this on Windows (with Chrome) and get two errors when selecting a cover image.

image

image

**Comment by mariotaku** [comment]
Sorry, but I can't reproduce it. Could you show the request body to this endpoint?

**Comment by ReenigneArcher** [comment]
Request Payload
```json
{
    "key": "igdb_10032",
    "url": "https://images.igdb.com/igdb/image/upload/t_cover_big_2x/co1rg8.png"
}
```

Response
```json
{
    "error": "Failed to download cover"
}
```

I don't know if it matters, but I am in incognito mode (I always use it for testing different builds)

**Comment by ReenigneArcher** [comment]
I have another problem running the portable version of this build (on my non development PC). `libcurl` is missing. Must not be an issue on my main laptop because it's installed somewhere on PATH.

Windows 10:
image

Sunshine will not start at all.

**Comment by ReenigneArcher** [comment]
While on topic of curl, probably needs to be added to these other locations:

| Pacakge Name | Location |
|--------------|----------|
| curl         | https://github.com/LizardByte/Sunshine/blob/b11d73599448fd4608c841811896d49c9dfa2236/.github/workflows/CI.yml#L440 |
| curl         | https://github.com/LizardByte/Sunshine/blob/b11d73599448fd4608c841811896d49c9dfa2236/packaging/macos/Portfile#L35 |
| curl         | https://github.com/LizardByte/Sunshine/blob/b11d73599448fd4608c841811896d49c9dfa2236/packaging/linux/aur/PKGBUILD#L12 |
| mingw-w64-curl | https://github.com/LizardByte/Sunshine/blob/b11d73599448fd4608c841811896d49c9dfa2236/.github/workflows/CI.yml#L726 |

**Comment by ReenigneArcher** [comment]
Still have this error with latest build: https://github.com/LizardByte/Sunshine/pull/216#issuecomment-1208271698

File: sunshine/httpcommon.cpp
Comment (ReenigneArcher) [inline sunshine/httpcommon.cpp:203]: Revisiting the original error about not being able to upload the cover.

Terminal output
```txt
[2022:08:20:18:36:42]: Error: Couldn't download [https://images.igdb.com/igdb/image/upload/t_cover_big_2x/co1rg8.png, code:77]
```

The `/assets/covers/` directory is created, and files are created inside the directory, but they are of 0 size.

image

I looked up code 77 and found this:
```
CURLE_SSL_CACERT_BADFILE (77)

Problem with reading the SSL CA cert (path? access rights?) 
```

I suppose it's an issue with the sunshine self signed certificate.

File: sunshine/confighttp.cpp
Comment (ReenigneArcher) [inline sunshine/confighttp.cpp:456]: Can we use this?

```cpp
const std::string coverdir = platf::appdata().string() + "/covers/"
```

Since the `SUNSHINE_ASSETS_DIR` will get overwritten on updates normally.

Comment (mariotaku) [reply sunshine/confighttp.cpp:456]: This sounds more reasonable.

File: sunshine/httpcommon.cpp
Comment (mariotaku) [reply sunshine/httpcommon.cpp:203]: I think this has nothing to do with self-signed certificate. But something is wrong with the certificate storage in your system.

Comment (ReenigneArcher) [reply sunshine/httpcommon.cpp:203]: Okay, I have no idea what cert curl is complaining about or where it expects to find it.

I'm testing in Windows. Sunshine does not store certs in the windows certificate store (or whatever it's called), just inside the install directory.

I get no warnings when navigating to the image url, only when navigating to sunshine's ui, due to the cert being self signed. You have to accept a warning in the browser to proceed. So figured it was related to that.

Comment (ReenigneArcher) [reply sunshine/httpcommon.cpp:203]: @mariotaku are you using a fresh instance of sunshine when testing, and do you use the certificate that is created by sunshine?

This is all I have tried.

- Reset certs in Firefox (issue is not browser dependent, was using Chrome incognito previously)
- Added sunshine cert pem file to Windows "Trusted Root Certification Authorities"
- Tried from a browser on another device instead of the host
- Was also going to test from Windows 10 but cannot start this build of sunshine on either of my Windows 10 systems (updated comment above).

Not really sure what else to try on my end.

Comment (mariotaku) [reply sunshine/httpcommon.cpp:203]: According to some docs, that's because openssl needs to be included, too. I'll take a look when I got time.

**Comment by ReenigneArcher** [comment]
Any update on this? Anything I can provide to help resolve the issues?

- [ ] Windows 10 - libcurl-4.dll missing - https://github.com/LizardByte/Sunshine/pull/216#issuecomment-1208271698
- [ ] Windows 11 and maybe others - `CURLE_SSL_CACERT_BADFILE (77)` error when trying to save image

**Comment by mariotaku** [comment]
Hi. Sorry I've gotten busy these days. Is it possible to ask you to make some changes if you have clues to fix the above issues?

**Comment by ReenigneArcher** [comment]
> Hi. Sorry I've gotten busy these days. Is it possible to ask you to make some changes if you have clues to fix the above issues?

I will see what I can do, as well as check with some other developers.

These are some references for the missing `dll` issue.
- https://stackoverflow.com/a/29670866/11214013
- https://stackoverflow.com/a/11545568/11214013
- https://stackoverflow.com/a/32476811/11214013

It appears that we need to place the dll next to `sunshine.exe` or statically linking curl.

**Comment by psyke83** [comment]
Hi,

It seems there was a problem with static linking that affected libcurl.
If you cherry-pick this commit, it should work ok: https://github.com/psyke83/SunshineStream/commit/92b90cd03325ae5fd07d593b3475e4e621163d8e

With the above commit:
```
# ldd sunshine.exe
        ntdll.dll => /c/Windows/SYSTEM32/ntdll.dll (0x7ff87ecf0000)
        KERNEL32.DLL => /c/Windows/System32/KERNEL32.DLL (0x7ff87db50000)
        KERNELBASE.dll => /c/Windows/System32/KERNELBASE.dll (0x7ff87c3f0000)
        ADVAPI32.dll => /c/Windows/System32/ADVAPI32.dll (0x7ff87e250000)
        msvcrt.dll => /c/Windows/System32/msvcrt.dll (0x7ff87da10000)
        sechost.dll => /c/Windows/System32/sechost.dll (0x7ff87dc70000)
        RPCRT4.dll => /c/Windows/System32/RPCRT4.dll (0x7ff87e050000)
        bcrypt.dll => /c/Windows/System32/bcrypt.dll (0x7ff87c6d0000)
        CRYPT32.dll => /c/Windows/System32/CRYPT32.dll (0x7ff87cbe0000)
        ucrtbase.dll => /c/Windows/System32/ucrtbase.dll (0x7ff87cae0000)
        ole32.dll => /c/Windows/System32/ole32.dll (0x7ff87de90000)
        d3d11.dll => /c/Windows/SYSTEM32/d3d11.dll (0x7ff878100000)
        D3DCOMPILER_47.dll => /c/Windows/SYSTEM32/D3DCOMPILER_47.dll (0x7ff878370000)
        combase.dll => /c/Windows/System32/combase.dll (0x7ff87e300000)
        dwmapi.dll => /c/Windows/SYSTEM32/dwmapi.dll (0x7ff87a0f0000)
        win32u.dll => /c/Windows/System32/win32u.dll (0x7ff87c910000)
        win32u.dll => /c/Windows/System32/win32u.dll (0x1cb6d9e0000)
        GDI32.dll => /c/Windows/System32/GDI32.dll (0x7ff87d1c0000)
        dxgi.dll => /c/Windows/SYSTEM32/dxgi.dll (0x7ff87acb0000)
        USER32.dll => /c/Windows/System32/USER32.dll (0x7ff87eb10000)
        gdi32full.dll => /c/Windows/System32/gdi32full.dll (0x7ff87c9d0000)
        IPHLPAPI.DLL => /c/Windows/SYSTEM32/IPHLPAPI.DLL (0x7ff87b7c0000)
        msvcp_win.dll => /c/Windows/System32/msvcp_win.dll (0x7ff87c7c0000)
        CRYPTSP.dll => /c/Windows/SYSTEM32/CRYPTSP.dll (0x7ff87bda0000)
        SETUPAPI.dll => /c/Windows/System32/SETUPAPI.dll (0x7ff87e660000)
        cfgmgr32.dll => /c/Windows/System32/cfgmgr32.dll (0x7ff87c770000)
        WLDAP32.dll => /c/Windows/System32/WLDAP32.dll (0x7ff87e1e0000)
        WINMM.dll => /c/Windows/SYSTEM32/WINMM.dll (0x7ff872930000)
        WS2_32.dll => /c/Windows/System32/WS2_32.dll (0x7ff87d9a0000)
        WSOCK32.dll => /c/Windows/SYSTEM32/WSOCK32.dll (0x7ff86dc00000)
        MSWSOCK.DLL => /c/Windows/SYSTEM32/MSWSOCK.DLL (0x7ff87bb40000)
```

I did a quick test and the cover finder feature seems to retrieve images correctly on Windows 10.

**Comment by ReenigneArcher** [comment]
@psyke83 Thank you so much!

@mariotaku looks like a couple of failures, should be easy to resolve.

1. clang reports an extra space
2. docker build failing... missing a dependency per https://github.com/LizardByte/Sunshine/actions/runs/3488506314/jobs/5837582884#step:9:3163
```txt
#15 1.895 CMake Error at /usr/share/cmake-3.22/Modules/FindPkgConfig.cmake:603 (message):
#15 1.895   A required package was not found
```
Should just be a matter of adding the dependency here: https://github.com/mariotaku/Sunshine/blob/6806ae2335565f61b1629502d23ceed3eb2c004f/Dockerfile#L17

`libcurl4-openssl-dev` if I'm not mistaken.

**Comment by mariotaku** [comment]
@ReenigneArcher Oh! I thought 22.04 is focal.

**Comment by ReenigneArcher** [comment]
Thanks for working on this everyone! I will test it in Windows 11 and 10 tonight. If it's all working, I'll assume we're good to merge, since it worked in Linux before.

**Comment by ReenigneArcher** [comment]
Unfortunately, with Windows 11 I am still not able to upload the cover.

image


stack trace:
```
useCover @ https://localhost:47990/apps:446
click @ :3
invokeWithErrorHandling @ https://localhost:47990/third_party/vue.js:1863
invoker @ https://localhost:47990/third_party/vue.js:2188
original._wrapper @ https://localhost:47990/third_party/vue.js:7547
```

**Comment by psyke83** [comment]
I saw this on Sunshine's log output: 
```
[2022:11:18:01:51:29]: Error: Couldn't download [https://images.igdb.com/igdb/image/upload/t_cover_big_2x/co4rjc.png, code:77]```

Code 77 is related to the certificate store; presumably curl is not able to find it in the expected format on a Windows system. I feel that instead of copying the certificate store, it would be better to use Window's own internal store instead.

To that effect, changing the build system to install `mingw-w64-x86_64-curl-winssl` (instead of `mingw-w64-x86_64-curl`, as they are in conflict), it seems that it will work. I might have had the `-winssl` varient installed when I tested, which is why I didn't catch the problem.

**Comment by psyke83** [comment]
Just to add some more context:

The `winssl` library seems to be included on Windows 10 or later only, so older versions may not work. Another alternative is to configure OpenSSL to use the Windows native store during compile time; see: https://curl.se/libcurl/c/CURLOPT_SSL_OPTIONS.html with the `CURLSSLOPT_NATIVE_CA` bitmask.

That might be worth investigating, but I'm not sure if it's worth the trouble? Windows 8.1 will be EOL in January 2023, and you'd have to be crazy to run Windows 7 in a networked environment.

**Comment by ReenigneArcher** [comment]
> The `winssl` library seems to be included on Windows 10 or later only, so older versions may not work.

Probably fine since Sunshine won't work in Windows 7 anyway. I've also yet to see a user report an issue and indicate they are using Windows 8.

**Comment by psyke83** [comment]
Well, either way seems to work on my system. Here's the OpenSSL workaround:
```
diff --git a/src/httpcommon.cpp b/src/httpcommon.cpp
index 092c2dc..e4747e3 100644
--- a/src/httpcommon.cpp
+++ b/src/httpcommon.cpp
@@ -197,6 +197,9 @@ bool download_file(const std::string &url, const std::string &file) {
   curl_easy_setopt(curl, CURLOPT_URL, url.c_str());
   curl_easy_setopt(curl, CURLOPT_WRITEFUNCTION, fwrite);
   curl_easy_setopt(curl, CURLOPT_WRITEDATA, fp);
+#ifdef _WIN32
+  curl_easy_setopt(curl, CURLOPT_SSL_OPTIONS, CURLSSLOPT_NATIVE_CA);
+#endif
   CURLcode result = curl_easy_perform(curl);
   if(result != CURLE_OK) {
     BOOST_LOG(error) << "Couldn't download ["sv << url << ", code:" << result << ']';

```

**Comment by ReenigneArcher** [comment]
The workaround you provided seems logical to me. Let's go with that direction.

**Comment by mariotaku** [comment]
I've applied @psyke83's patch. Really appreciate your suggestions!

**Review by ReenigneArcher — APPROVED** [review]

**Pull request merged**
